ponycheck/generator.pony

```diff
@@ -1,68 +1,275 @@
+
+    If generating an example value is costly, it might be more efficient
+    to simply return the generated value and only shrink in big steps or do no
+    shrinking at all.
+    If generating values is lightweight, shrunken values should also be returned.
+    """
     _gen.generate(rnd)
 
-  fun shrink(t: T): (T^, Seq[T]) =>
+  fun shrink(t: T): ValueAndShrink[T] =>
+    """
+    Simplify the given value.
+
+    As the returned value can also be `iso`, it needs to be consumed and returned
+
+    It is preffered to already return a `ValueAndShrink` from `generate`.
+    """
     _gen.shrink(consume t)
 
+  fun generate_value(rnd: Randomness): T^ =>
+    _gen.generate_value(rnd)
+
+  fun generate_and_shrink(rnd: Randomness): ValueAndShrink[T] =>
+    _gen.generate_and_shrink(rnd)
+
   fun filter(predicate: {(T): (T^, Bool)} box): Generator[T] =>
+    """
+    apply `predicate` to the values generated by this Generator
+    and only values for which `predicate` returns `true`.
+
+    Example:
+
+    ```pony
+    let even_i32s =
+      Generators.i32()
+        .filter(
+          {(t: I32}: (I32^, Bool) => (t % 2) == 0 })
+    ```
+    """
     Generator[T](
       object is GenObj[T]
-        fun generate(rnd: Randomness): T^ =>
-          (var t, var matches) = predicate(_gen.generate(rnd))
-          while not matches do
-            (t, matches) = predicate(_gen.generate(rnd))
+        fun generate(rnd: Randomness): GenerateResult[T] =>
+          (let t: T, let shrunken: Iterator[T^]) = _gen.generate_and_shrink(rnd)
+          (let t1, let matches) = predicate(consume t)
+          if not matches then
+            generate(rnd) // recurse, this might recurse infinitely
+          else
+            // filter the shrunken examples
+            (consume t1, _filter_shrunken(shrunken))
           end
-          consume t
-        // TODO: shrink by calling _gen.shrink and filter the returned list
-        // using predicate
+
+        fun shrink(t: T): ValueAndShrink[T] =>
+          """
+          shrink `t` using the generator this one filters upon
+          and call the filter predicate on the shrunken values
+          """
+          (let s, let shrunken: Iterator[T^]) = _gen.shrink(consume t)
+          (consume s, _filter_shrunken(shrunken))
+
+        fun _filter_shrunken(shrunken: Iterator[T^]): Iterator[T^] =>
+          Iter[T^](shrunken)
+            .filter_map[T^]({
+              (t: T): (T^| None) =>
+                match predicate(consume t)
+                | (let matching: T^, true) => consume matching
+                end
+            })
       end)
 
-  fun map[U](fn: {(T): U^} box): Generator[U] =>
+  fun map[U](fn: {(T): U^} box)
+    : Generator[U]
+  =>
+    """
+    apply function `fn` to each value of this iterator
+    and yield the results.
+
+    Example:
+
+    ```pony
+    let single_code_point_string_gen =
+      Generators.u32()
+        .map[String]({(u: U32): String => String.from_utf32(u) })
```

Comment:
The examples with simple lambdas like this should probably use the new lambda syntax.
'''pony
.map[String]({(u) => String.from_utf32(u) })
'''

Reply:
Right, looks much nicer.
I adapted it everywhere possible.